File: labs/l5/multilingual_pattern_lab/go/verifiable_action.go
package main

import "strings"

const (
	safetyRule       = "delete_all"
	readyReason      = "ready_for_execution"
	blockedReason    = "blocked_by_safety_rule"
)

type PlanStep struct {
	Intent        string
	Tool          string
	RollbackHint  string
}

type PatternResult struct {
	Steps      []PlanStep
	StopReason string
}

func readyPlan() PatternResult {
	return PatternResult{
		Steps: []PlanStep{
			{"clarify", "validator", "remove unclear fields"},
			{"execute", "tool_gateway", "restore previous state"},
			{"verify", "eval_probe", "disable path"},
		},
		StopReason: readyReason,
	}
}

func plan(request string, safetyRules []string) PatternResult {
	for _, rule := range safetyRules {
		if strings.Contains(request, rule) {
			return PatternResult{StopReason: blockedReason}
		}
	}
	return readyPlan()
}

func main() {
	rules := []string{safetyRule}
	if plan("update profile", rules).StopReason != readyReason {
		panic("expected ready plan")
	}
	if len(plan("delete_all", rules).Steps) != 0 {
		panic("expected blocked plan")
	}
}

File: labs/l5/multilingual_pattern_lab/node/verifiable_action.mjs
export const SAFETY_RULE = "delete_all";
export const READY_REASON = "ready_for_execution";
export const BLOCKED_REASON = "blocked_by_safety_rule";

const readyPlan = [
  { intent: "clarify", tool: "validator", rollbackHint: "remove unclear fields" },
  { intent: "execute", tool: "tool_gateway", rollbackHint: "restore previous state" },
  { intent: "verify", tool: "eval_probe", rollbackHint: "disable path" },
];

export function plan(request, safetyRules = [SAFETY_RULE]) {
  if (safetyRules.some((rule) => request.includes(rule))) {
    return { steps: [], stopReason: BLOCKED_REASON };
  }
  return { steps: readyPlan, stopReason: READY_REASON };
}

if (process.argv[1] && import.meta.url === `file://${process.argv[1]}`) {
  console.log(JSON.stringify(plan("update profile")));
  console.log(JSON.stringify(plan("delete_all")));
}
